runtime/lua/vim/lsp.lua

```diff
@@ -1159,6 +1168,7 @@ function lsp.buf_attach_client(bufnr, client_id)
       on_reload = function()
         local params = { textDocument = { uri = uri; } }
         for_each_buffer_client(bufnr, function(client, _)
+          changetracking.reset_buf(client, bufnr)
```

Comment:
Shouldn't this flush instead of throw it away?

Reply:
`client.notify` doesn't cause a flush

Reply:
That can be fixed too, but I think that explicitly performing a reset here won't hurt

Reply:
I'm aware that notify doesn't flush. It can't flush as long as the debounce itself calls `client.notify` to do the debouncing.

My question is whether this here should be calling flush instead of reset_buf.

Reply:
Well, currently calling `flush` makes more sense because `reset_buf` currently doesn't delete the entry in `pending_changes` for the buffer, but that is beside the point. The thing is that we are sending new buffer contents immediately afterwards, and with the pending didChange events meant to be applied to the version of the buffer before reload, they are basically useless because their results will be discarded as soon as the client processes didOpen. Thus I think that it is better to throw them away.

Reply:
There's no point to flushing as we send the full document here https://github.com/mjlbach/neovim/blob/988fd1d825c99e6212f25c10236eda528a89dda1/runtime/lua/vim/lsp.lua#L478-L486

I'll modify reset_buf to flush pending changes (I thought we overwriting state entirely in change_tracking.init

Reply:
I think `debounce` should only debounce - not additionally optimize away change sets, even if they're most likely unnecessary. I'm pretty sure the optimization wouldn't have any practical impact.

And I think keeping the semantics of debounce on/off close to each other helps reasoning about potential issues or with trouble shooting. Both should result in the same set of changes getting eventually reported to the server (although in a different format due the batching)